### PR TITLE
[fix]:[CDS-75190] adding fix for service type based on workflow type

### PR DESCRIPTION
**Pull request by RajBaviskar**
## Describe your changes

- adding a fix for service type based on workflow type

https://github.com/harness/harness-core/assets/7086440/ff279eea-4889-4524-abdb-b92d5f0fc631

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51345)

### CODE REVIEW

**Comment by RajBaviskar** [comment]
trigger uts

**Comment by RajBaviskar** [comment]
trigger uts

File: 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java [hunk 1]
```diff
@@ -4328,6 +4328,19 @@ public List<String> getLastSuccessfulWorkflowExecutionIds(String appId, String w
...
+    }
+    return wingsPersistence.createQuery(WorkflowExecution.class)
+        .filter(WorkflowExecutionKeys.accountId, accountId)
+        .filter(WorkflowExecutionKeys.appId, appId)
+        .filter(WorkflowExecutionKeys.serviceIds, serviceId)
+        .filter(WorkflowExecutionKeys.status, SUCCESS)
+        .order(Sort.descending(WorkflowExecutionKeys.createdAt))
+        .get();
+  }
+
```
Comment (lasdox) [inline 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java:4343]: we already have same method in WorkflowExecutionService.
Could you please check if you can use that instead?

Comment (lasdox) [reply 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java:4343]: Also shouldnt the WorkflowExecution be filtered by this workflowId?

Comment (PragyeshMishra01) [reply 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java:4343]: Agreed, we should try to use existing method.

**Review by PragyeshMishra01 — COMMENTED** [review]
added a comment

File: 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java (hunk 1 above)
Comment (RajBaviskar) [reply 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java:4343]: one from WorkflowExecutionService requires the workflowId. The one added we need to fetch the last successful execution of a workflow for a given service.

**Comment by RajBaviskar** [comment]
trigger uts

File: 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java (hunk 1 above)
Comment (lasdox) [reply 400-rest/src/main/java/software/wings/service/impl/workflow/WorkflowServiceImpl.java:4343]: why not just giving workflowId? Im not seeing why we need to create it in WorkflowService

**Comment by RajBaviskar** [comment]
trigger ut0

**Review by lasdox — APPROVED** [review]
lGTM

**Review by Fernando-Dourado — APPROVED** [review]

**Review by arvind-choudhary-h — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 14 of 15 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51345)* status: 17 file(s) to review. 1. unresolved discussions.

**Comment by arvind-choudhary-h** [comment]
trigger review

**Pull request merged**